UIIN-2397: Item Create/Edit screens: Replace custom RepeatableField with component from stripes

mariia-aloshyna opened the pull request:
<!--
  If you have a relevant JIRA issue number, please put it in the issue title.
  Example: UIIN-817 Orders schema updates
-->

## Purpose
<!--
  Why are you making this change? There is nothing more important
  to provide to the reviewer and to future readers than the cause
  that gave rise to this pull request. Be careful to avoid circular
  statements like "the purpose is to update the schema." and
  instead provide an explanation like "there is more data to be provided and stored for Purchase Orders
  which is currently missing in the schema"

  The purpose may seem self-evident to you now, but the standard to
  hold yourself to should be "can a developer parachuting into this
  project reconstruct the necessary context merely by reading this
  section."

  If you have a relevant JIRA issue, add a link directly to the issue URL here.
  Example: https://issues.folio.org/browse/UIIN-817
 -->
The custom RepeatableField component was implemented a while ago and causes style issues, replacing it with the RepeatableField component from stripes/components fixes the bugs and makes the Inventory consistent with the whole application.

## Approach
<!--
 How does this change fulfill the purpose? It's best to talk
 high-level strategy and avoid code-splaining the commit history.

 The goal is not only to explain what you did, but help other
 developers *work* with your solution in the future.
-->
- Replace custom RepeatableField with RepaetableField from stripes 
- Get rid of custom RepeatableField component
- Fix code smells
- Add unit tests

#### TODOS and Open Questions
<!-- OPTIONAL
- [ ] Use GitHub checklists. When solved, check the box and explain the answer.
-->

## Learning
<!-- OPTIONAL
  Help out not only your reviewer, but also your fellow developer!
  Sometimes there are key pieces of information that you used to come up
  with your solution. Don't let all that hard work go to waste! A
  pull request is a *perfect opportunity to share the learning that
  you did. Add links to blog posts, patterns, libraries or addons used
  to solve this problem.
-->

## Pre-Merge Checklist
Before merging this PR, please go through the following list and take appropriate actions.
- [ ] I've added appropriate record to the CHANGELOG.md
- Does this PR meet or exceed the expected quality standards?
  - [ ] Code coverage on new code is 80% or greater
  - [ ] Duplications on new code is 3% or less
  - [ ] There are no major code smells or security issues
- Does this introduce breaking changes?
  - [ ] If any API-related changes - okapi interfaces and permissions are reviewed/changed correspondingly
  - [ ] There are no breaking changes in this PR.

If there are breaking changes, please **STOP** and consider the following:

- What other modules will these changes impact?
- Do JIRAs exist to update the impacted modules?
  - [ ] If not, please create them
  - [ ] Do they contain the appropriate level of detail?  Which endpoints/schemas changed, etc.
  - [ ] Do they have all they appropriate links to blocked/related issues?
- Are the JIRAs under active development?
  - [ ] If not, contact the project's PO and make sure they're aware of the urgency.
- Do PRs exist for these changes?
  - [ ] If so, have they been approved?

Ideally all of the PRs involved in breaking changes would be merged in the same day to avoid breaking the folio-testing environment.  Communication is paramount if that is to be achieved, especially as the number of intermodule and inter-team dependencies increase.

While it's helpful for reviewers to help identify potential problems, ensuring that it's safe to merge is ultimately the responsibility of the PR assignee.

maccabeelevine commented:
Hi @mariia-aloshyna, it's great that this PR is happening.  I implemented the bound-with functionality over the past few months and noticed how much the ui-inventory version of RepeatableField had diverged from the stripes component.  I didn't attempt this refactoring at the time because it seemed like there were intentional UX differences between the two versions, such as a) whether the fields should be forced full-width or just left aligned, b) whether the field labels should be repeated above each row.  It's a good idea to finally make those decisions and standardize it.

Unfortunately for that bound-with work I added even more differences to the ui-inventory version for required functionality, like the canDelete functionality being specific to a each row.  Sorry you have to migrate that as well to stripes-components.  There's also the extra data loaded by BoundWithFieldRow.  

If you have any questions about the bound-with stuff as you finish this, I'm happy to help.

mariia-aloshyna commented:
> Hi @mariia-aloshyna, it's great that this PR is happening. I implemented the bound-with functionality over the past few months and noticed how much the ui-inventory version of RepeatableField had diverged from the stripes component. I didn't attempt this refactoring at the time because it seemed like there were intentional UX differences between the two versions, such as a) whether the fields should be forced full-width or just left aligned, b) whether the field labels should be repeated above each row. It's a good idea to finally make those decisions and standardize it.
> 
> Unfortunately for that bound-with work I added even more differences to the ui-inventory version for required functionality, like the canDelete functionality being specific to a each row. Sorry you have to migrate that as well to stripes-components. There's also the extra data loaded by BoundWithFieldRow.
> 
> If you have any questions about the bound-with stuff as you finish this, I'm happy to help.

Hi @maccabeelevine Thank you for clarification! I refactored BoundWithTitles a bit, could you please check locally if it works as expected before it's merged?

maccabeelevine commented on src/edit/items/ItemForm.js:
```diff
@@ -922,51 +820,9 @@ class ItemForm extends React.Component {
-                      name="boundWithTitles"
-                      label={<FormattedMessage id="ui-inventory.boundWithTitles" />}
-                      canAdd={false}
-                      canDelete={(fields, fieldIndex) => {
```
Screenshot of what it should do from editing this on snapshot: https://folio-snapshot.dev.folio.org/inventory/edit/ce9dd893-c812-49d5-8973-d55d018894c4/9e8dc8ce-68f3-4e75-8479-d548ce521157/f4b8c3d1-f461-4551-aa7b-5f45e64f236c?query=rapport&sort=title

![image](https://user-images.githubusercontent.com/31278545/236258561-d027b58a-347a-40be-b18c-43d629d4f70b.png)

maccabeelevine reviewed (changes requested):
Overall your changes to my bound-with code are a big improvement, thank you!  The alignment of the trash-can icon is also great.

I commented on a few specific problems I'm seeing.

mariia-aloshyna commented:
Good catches @maccabeelevine! Everything is fixed now

maccabeelevine reviewed (approved):
The two issues I identified before are fixed.  Everything looks good to me!
